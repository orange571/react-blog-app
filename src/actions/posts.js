import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment'

// ADD_POST
export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      isReadable = false,
      author = '',
    } = postData;
    const post = { title, body, createdAt, isReadable, author, uid };

    return database.ref(`posts`).push(post).then((ref) => {
      const postId = ref.key;
      database.ref(`users/${uid}/posts`).update({[postId] : isReadable}).then(()=>{
        if(isReadable){
          database.ref(`isPublic`).push({id: postId})
        };
      });
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

// REMOVE_POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).remove().then(() => {
      //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
          const wasPublic = snapshot.val();
          if (!!wasPublic){
            const publicPostId = Object.keys(wasPublic)[0];
            database.ref(`isPublic/${publicPostId}`).remove().then(() => {
              //REMOVE FROM USER List
              database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
              .catch(error => console.log(error));
            }).catch(error => console.log(error));
          } else {
            //REMOVE FROM USER List
            database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
            .catch(error => console.log(error));
          }
      });
      dispatch(removePost({ id }));
    });
  };
};

// EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).update(updates).then(() => {
      //ADD TO PUBLIC LIST IF WAS NOT THERE ALREADY
      if(updates.isReadable){
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            if (!wasPublic){
              database.ref(`isPublic`).push({id: id})
            }
        });
        //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      } else {
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            const publicPostId = Object.keys(wasPublic)[0];
            if (!!wasPublic){
              database.ref(`isPublic/${publicPostId}`).remove();
            }
        });
      }
      database.ref(`users/${uid}/posts`).update({[id]: updates.isReadable});
      dispatch(editPost(id, updates));
    });
  };
};

// SET_POSTS
export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      if(!snapshot.val()) {
        let posts = [];
        dispatch(setPosts(posts));
      } else {
        return new Promise((resolve, reject) => {
          const postIds = [];
          snapshot.forEach((childSnapshot) => {
            postIds.push(childSnapshot.key);
          });
          let posts = [];
          postIds.forEach((postId, index)=>{
            database.ref(`posts/${postId}`).once('value').then((postSnapshot) => {
              posts.push({
                id: postSnapshot.key,
                ...postSnapshot.val()
              });
              if (index === postIds.length-1) {
                resolve(posts);
              }
            }).catch((error) => {
              console.log("error", error);
              reject(error);
            });
          });
        }).then((posts)=>{
          dispatch(setPosts(posts));
        }).catch(error => {
          console.log(error)
        });
      }
    }).catch((error)=>{
      console.log('error', error);
    });
  };
};
