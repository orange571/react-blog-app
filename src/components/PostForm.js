import React from 'react';
import moment from 'moment';

export default class PostForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      title: props.post ? props.post.title : '',
      body: props.post ? props.post.body : '',
      createdAt: props.post ? moment(props.post.createdAt) : moment(),
      author: props.post ? props.post.author: props.displayName,
      error: ''
    };
  }
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };
  onBodyChange = (e) => {
    const body = e.target.value;
    this.setState(() => ({ body }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.title || !this.state.body) {
      this.setState(() => ({ error: 'Please provide a title and post body.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        title: this.state.title,
        createdAt: this.state.createdAt.valueOf(),
        body: this.state.body,
        author: this.state.author.valueOf(),
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Title"
          autoFocus
          className="text-input"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea
          placeholder="Add a body for your post"
          className="textarea"
          value={this.state.body}
          onChange={this.onBodyChange}
        >
        </textarea>
        <div className="form__button-container">
          <button type="submit" className="button">Save Post</button>
          {this.props.onRemove && (<button className="button button--secondary" onClick={this.props.onRemove}>Remove Post</button>)}
        </div>
      </form>
    )
  }
}
