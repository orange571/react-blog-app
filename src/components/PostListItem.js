import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const PostListItem = ({ id, title, body, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <h3 className="list-item__title list-item__title--private">{title}</h3>
    <h3 className="list-item__data">{moment(createdAt).format('M/D/YY')}</h3>
  </Link>
);

export default PostListItem;
