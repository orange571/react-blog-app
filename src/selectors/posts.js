import moment from 'moment';

// Get visible posts

export default (posts, { text, sortBy }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title.localeCompare(b.title);
    }
  });
};
