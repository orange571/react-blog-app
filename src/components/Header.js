import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogin, startLogout } from '../actions/auth';

export const Header = ({ startLogout, startLogin, displayName }) => (
  <header className="header">
    <div className="show-for-desktop content-container content-container--header">
      <div className="header__content">
          {
            displayName ? (
              <div  className="header__left-nav">
                <Link className="header__title" to="/dashboard">
                  <h1>Thought Collector</h1>
                </Link>
                <NavLink to="/dashboard" className="header__link" activeClassName="is-active" exact={true}>Dashboard</NavLink>
                <NavLink to="/public" className="header__link "activeClassName="is-active">Browse</NavLink>
              </div>
            ) : (
              <div  className="header__left-nav">
                <Link className="header__title" to="/dashboard">
                  <h1>Thought Collector</h1>
                </Link>
                <Link className="header__link" to="/public">
                  Browse
                </Link>
              </div>
            )
          }
        {
          displayName ? (
            <div className="header__right-nav">
              <div className="header__username">Signed in as {displayName}</div>
              <button className="button button--link logout" onClick={startLogout}>Logout</button>
            </div>
          ) : (
            <div className="header__right-nav">
              <button className="button login" onClick={startLogin}>Login</button>
            </div>
          )
        }
      </div>
    </div>
    <div className="show-for-mobile">
      <div  className="header_content--top">
        <Link className="header__title" to="/dashboard">
          <h1>Thought Collector</h1>
        </Link>
        {displayName && <div className="header__username">Signed in as {displayName}</div>}
      </div>
      {
        displayName ? (
          <div className="header_content--bottom">
            <div className="content-container--bottom-nav">
              <NavLink to="/dashboard" className="header__link" activeClassName="is-active" exact={true}>Dashboard</NavLink>
              <NavLink to="/public" className="header__link "activeClassName="is-active">Browse</NavLink>
              <button className="button button--link logout" onClick={startLogout}>Logout</button>
            </div>
          </div>
        ) : (
          <div className="header_content--bottom">
            <div className="content-container--bottom-nav">
              <Link className="header__link" to="/public">Browse</Link>
              <button className="button button--link login" onClick={startLogin}>Login</button>
            </div>
          </div>
        )
      }
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
  startLogin: () => dispatch(startLogin()),
});


const mapStateToProps = (state) => {
  return {
    displayName: state.auth.displayName
  };
};


export default connect(mapStateToProps, mapDispatchToProps)(Header);
