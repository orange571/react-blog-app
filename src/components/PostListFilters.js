import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setTextFilter, sortByDate, sortByTitle} from '../actions/filters';

export class PostListFilters extends React.Component {
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'title') {
      this.props.sortByTitle();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__left">
            <div className="input-group__item">
              <input
                type="text"
                className="text-input"
                placeholder="Search posts"
                value={this.props.filters.text}
                onChange={this.onTextChange}
              />
            </div>
            <div className="input-group__item">
              <select
                className="select"
                value={this.props.filters.sortBy}
                onChange={this.onSortChange}
              >
                <option value="title">Title</option>
                <option value="date">By Date</option>
              </select>
            </div>
          </div>  
          <div className="input-group__item">
                <Link className="button" to="/create">Add Post</Link>
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByTitle: () => dispatch(sortByTitle()),
});

export default connect(mapStateToProps, mapDispatchToProps)(PostListFilters);
