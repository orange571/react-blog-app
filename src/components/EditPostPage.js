import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startEditPost, startRemovePost } from '../actions/posts';

export class EditPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isReadable : props.post ? props.post.isReadable : false,
    };
  }
  onSubmit = (post) => {
    this.props.startEditPost(this.props.post.id, post);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemovePost({ id: this.props.post.id });
    this.props.history.push('/');
  };
  displayLink = (isReadable) => {
    this.setState(()=>({isReadable:isReadable}));
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Post</h1>
            {this.state.isReadable ? (
                <div id="link-container">
                  <div>Use the link below to share with others</div>
                  <a href={"/view/"+ this.props.match.params.id} target="_blank">https://thought-collector-app.herokuapp.com/view/{this.props.match.params.id}</a>
                </div>
              ) : (
                <div id="link-container">
                  <div>Get a link to share your post with others</div>
                  <div>when you check "Make Public"</div>
                </div>
              )
            }
          </div>
        </div>
        <div className="content-container">
          <PostForm
            post={this.props.post}
            onSubmit={this.onSubmit}
            onRemove={this.onRemove}
            displayLink={this.displayLink}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  post: state.posts.find((post) => post.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: (data) => dispatch(startRemovePost(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);
