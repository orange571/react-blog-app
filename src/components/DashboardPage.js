import React from 'react';
import PostList from './PostList';
import PostListFilters from './PostListFilters';
//import PostsSummary from './PostsSummary';

const DashboardPage = () => (
  <div>
    <PostListFilters />
    <PostList />    
  </div>
);

export default DashboardPage;
