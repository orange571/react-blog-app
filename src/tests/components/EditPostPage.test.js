import React from 'react';
import { shallow } from 'enzyme';
import posts from '../fixtures/posts';
import { EditPostPage } from '../../components/EditPostPage';

let startEditPost, startRemovePost, history, wrapper;

beforeEach(() => {
  startEditPost = jest.fn();
  startRemovePost = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditPostPage
      startEditPost={startEditPost}
      startRemovePost={startRemovePost}
      history={history}
      post={posts[2]}
    />
  );
});

test('should render EditPostPage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditPost', () => {
  wrapper.find('PostForm').prop('onSubmit')(posts[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditPost).toHaveBeenLastCalledWith(posts[2].id, posts[2]);
});

test('should handle startRemovePost', () => {
  wrapper.find('PostForm').prop('onRemove')();
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemovePost).toHaveBeenLastCalledWith({
    id: posts[2].id
  });
});
