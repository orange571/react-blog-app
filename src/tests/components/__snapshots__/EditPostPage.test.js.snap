// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPostPage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
      <div
        id="link-container"
      >
        <div>
          Get a link to share your post with others
        </div>
        <div>
          when you check "Make Public"
        </div>
      </div>
    </div>
  </div>
  <div
    className="content-container"
  >
    <PostForm
      displayLink={[Function]}
      onRemove={[Function]}
      onSubmit={[Function]}
      post={
        Object {
          "author": "thisismytestname",
          "body": "Credit cards are thin",
          "createdAt": 345600000,
          "id": "3",
          "isReadable": false,
          "title": "Credit Card",
          "uid": "thisismytestuid",
        }
      }
    />
  </div>
</div>
`;
